File: medium/207-Course-Schedule/TopologicalSort.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, Set<Integer>> graph = new HashMap<>();
        int[] inDegree = new int[numCourses];
        
        for (int i = 0; i < numCourses; i++) {
            graph.put(i, new HashSet<>());
        }
        
        for (int[] pair: prerequisites) {
            graph.get(pair[1]).add(pair[0]);
            inDegree[pair[0]] += 1;
        }
        
        List<Integer> order = new ArrayList<>();
        Queue<Integer> queue = new LinkedList<>();
        
        for (int i = 0; i < inDegree.length; i++) {
            if (inDegree[i] == 0) queue.add(i);
        }
        
        int node;
        
        while (!queue.isEmpty()) {
            node = queue.poll();
            order.add(node);
            if (graph.containsKey(node)) {
                for (int adj: graph.get(node)) {
                    if (--inDegree[adj] == 0) queue.offer(adj);
                }
            }
        }
        
        return order.size() == numCourses;   
    }
}

File: 207-Course-Schedule/TopologicalSort.kt
class Solution {
    fun canFinish(numCourses: Int, prerequisites: Array<IntArray>): Boolean {
        var graph = mutableMapOf<Int, MutableSet<Int>>()
        var indegree = Array<Int>(numCourses) {0}
        
        prerequisites.forEach {
            if (graph[it[1]] == null) graph[it[1]] = mutableSetOf<Int>()
            graph[it[1]]!!.add(it[0])
            indegree[it[0]] += 1
        }
        
        var order = mutableListOf<Int>()
        var queue = ArrayDeque<Int>()
        
        indegree.forEachIndexed {i, e -> if (e == 0) queue.add(i)}
        
        var node: Int
        
        while (queue.size > 0) {
            node = queue.removeFirst()
            order.add(node)
            graph[node]?.forEach {if (--indegree[it] == 0) queue.add(it)}
        }
        
        return order.size == numCourses
    }
}
